backend/de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/ddl/public/functions/getLocalTaxReportingConversionRateDate.sql

```diff
@@ -0,0 +1,96 @@
+/*
+ * #%L
+ * de.metas.adempiere.adempiere.migration-sql
+ * %%
+ * Copyright (C) 2023 metas GmbH
+ * %%
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation, either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program. If not, see
+ * <http://www.gnu.org/licenses/gpl-2.0.html>.
+ * #L%
+ */
+
+CREATE OR REPLACE FUNCTION getLocalTaxReportingConversionRateDate(p_taxReportRateBase varchar, p_c_calendar_id numeric, p_dateinvoiced timestamp with time zone, p_movementdate timestamp with time zone)
+    RETURNS timestamp without time zone
+    LANGUAGE plpgsql
+AS
+$$
+
+DECLARE
+
+    dateResult timestamp with time zone;
+    count numeric;
+    attempt int := 1;
+
+BEGIN
+    IF (p_taxReportRateBase IS NULL OR TRIM(p_taxReportRateBase) = '') THEN
+        RAISE EXCEPTION 'taxReportRateBase is not set';
+    END IF;
+
+    IF (p_c_calendar_id <= 0) THEN
+        RAISE EXCEPTION 'c_calendar_id is not set';
+    END IF;
+
+    IF (p_dateinvoiced IS NULL) THEN
+        RAISE EXCEPTION 'date is not set';
+    END IF;
+
+    -- I  - invoice date
+    IF (p_taxReportRateBase = 'I') THEN
+        dateResult = trunc(p_dateinvoiced, 'DD');
+        dateResult = dateResult - 1;
+    END IF;
+    -- S  - goods issue / shipment date
+    IF (p_taxReportRateBase = 'S') THEN
+        dateResult = trunc(p_movementdate, 'DD');
+        dateResult = dateResult - 1;
+    END IF;
+    -- BI - 1 day before invoice date
+    IF (p_taxReportRateBase = 'BI') THEN
+        dateResult = trunc(p_dateinvoiced, 'DD');
+        dateResult = dateResult - 1;
+    END IF;
+
+    WHILE (attempt < 10) LOOP
```

Comment:
i think something like this would by much more efficient:
```sql

SELECT
       '2022-03-03'::date - offset_days as dateresult,
       offset_days,
       nbd.name
FROM GENERATE_SERIES(0, 100) offset_days
         LEFT OUTER JOIN c_nonbusinessday nbd ON (
            c_calendar_id = 111/*p_c_calendar_id*/
        AND isactive = 'Y'
        AND TRUNC(date1, 'DD')::date = '2022-03-03'::date - offset_days
    )
WHERE nbd.c_nonbusinessday_id IS NULL
order by dateresult desc
limit 1

```

anyways IMHO we shall increase the checking period from 10 to something like 100.